tensorzero-core/src/providers/tensorzero_relay.rs

```diff
@@ -0,0 +1,571 @@
+) -> Result<(), Error> {
+    let res = inject_extra_request_data(
+        &request.request.extra_body,
+        &request.request.extra_headers,
+        model_provider,
+        request.model_name,
+        &mut serde_json::json!({}),
+    )?;
+    if res.matched_any_headers_or_body {
+        return Err(Error::new(ErrorDetails::Inference {
+            message: "Extra headers and body are not supported for `tensorzero_relay` provider"
+                .to_string(),
+        }));
+    }
+    Ok(())
+}
+
+impl TensorZeroRelayProvider {
+    pub async fn new(
+        gateway_base_url: Url,
+        function_name: Option<String>,
+        model_name: Option<String>,
+        variant_name: Option<String>,
+        credentials: TensorZeroRelayCredentials,
+        http_client: TensorzeroHttpClient,
+    ) -> Result<Self, Error> {
+        match credentials {
+            TensorZeroRelayCredentials::None => {}
+            _ => {
+                return Err(Error::new(ErrorDetails::Config {
+                    message: "Credentials are not yet supported for `tensorzero_relay` provider"
+                        .to_string(),
+                }))
+            }
+        }
+        // Validate that exactly one of function_name or model_name is set
+        let target_type = match (&function_name, &model_name) {
+            (Some(fn_name), None) => TensorZeroRelayTargetType::Function {
+                function_name: fn_name.clone(),
+                variant_name,
+            },
+            (None, Some(model)) => {
+                // Validate that variant_name is not set when using model_name
+                if variant_name.is_some() {
+                    return Err(Error::new(ErrorDetails::Config {
+                        message: "variant_name can only be set when function_name is provided"
+                            .to_string(),
+                    }));
+                }
+                TensorZeroRelayTargetType::Model {
+                    model_name: model.clone(),
+                }
+            }
+            (Some(_), Some(_)) => {
+                return Err(Error::new(ErrorDetails::Config {
+                    message: "Cannot provide both function_name and model_name".to_string(),
+                }))
+            }
+            (None, None) => {
+                return Err(Error::new(ErrorDetails::Config {
+                    message: "Exactly one of function_name or model_name must be provided"
+                        .to_string(),
+                }))
+            }
+        };
+
+        let client =
+            crate::client::ClientBuilder::new(crate::client::ClientBuilderMode::HTTPGateway {
+                url: gateway_base_url.clone(),
+            })
+            .with_http_client(http_client)
+            .build()
+            .await
+            .map_err(|e| {
+                Error::new(ErrorDetails::InternalError {
+                    message: e.to_string(),
+                })
+            })?;
```

Comment:
is this cheap enough to build given an http client that we're fine doing this on the hot path? If so, please add a comment marking that.